build: something is off with EXECGEN_TARGETS and logictest-bins

Here is a repro of some weird behavior of logic test binary:
```
git reset --hard 885b0ed7eb8a10ecb88566d5fb337c3a68fd1273

# This succeeds.
make clean && make testbaselogic FILES=vectorize_types

git reset --hard HEAD^

# This fails.
make testbaselogic FILES=vectorize_types
```
The failure is
```
EXECGEN pkg/sql/colexec/hashtable_distinct.eg.go
EXECGEN pkg/sql/colexec/hashtable_full.eg.go
ERROR: unrecognized filename: hashtable_full.eg.go
ERROR: unrecognized filename: hashtable_distinct.eg.go
make: *** [pkg/sql/colexec/hashtable_distinct.eg.go] Error 1
make: *** Waiting for unfinished jobs....
make: *** [pkg/sql/colexec/hashtable_full.eg.go] Error 1
```
Note that the commit in question introduced these two `EXECGEN_TARGETS` and removed another one (namely, it removed `hashtable.eg.go` and added `hashtable_full.eg.go` and `hashtable_distinct.eg.go`).

--- a/Makefile
+++ b/Makefile
@@ -1558,7 +1558,7 @@ endif
 # to the present, because then it will becomes newer
 # than all the other produced artifacts downstream and force
 # them to rebuild too.
-%.eg.go: bin/execgen
+$(EXECGEN_TARGETS): bin/execgen
 	@echo EXECGEN $@
 	@execgen $@ > $@.tmp || { rm -f $@.tmp; exit 1; }
 	@cmp $@.tmp $@ 2>/dev/null && rm -f $@.tmp || mv -f $@.tmp $@
@@ -1569,6 +1569,12 @@ endif
 	    target_timestamp_file=bin/execgen; \
 	  fi; \
 	  touch -r $$target_timestamp_file $@
+
+# Add a catch-all rule for any non-existent execgen generated
+# files. This prevents build errors when switching between branches
+# that introduce or remove execgen generated files as these files are
+# persisted in bin/%.d dependency lists (e.g. in bin/logictest.d).
+%.eg.go: ;
 
 optgen-defs := pkg/sql/opt/ops/*.opt
 optgen-norm-rules := pkg/sql/opt/norm/rules/*.opt

